rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+By example:
+
+-   `3.9.2` to `3.9.3`: always considered a bug fix
+-   `3.9.3` to `4.0.0`: *may or may not* introduce breaking changes
+-   `3.8.3` to `3.9.0`: *may or may not* introduce breaking changes
+
+### Defining breaking changes
+
+When making a change to the types of the public interface, it is subject to the same constraints as runtime code: *breaking the published types entails a breaking change.* Not all changes to published types are *breaking*, however: some changes will continue to allow user code to continue working without any issue. Thus, a *breaking* change to types is a change where previously-allowed code will stop type-checking.
+
+#### Breaking changes to type definitions
+
+A breaking change to a type definition occurs when—
+
+-   the type changes entirely, for example if a function previously accepted `number` and now accepts `{ count: number }`—since the user will have to change all calls to the function ([playground][changed-type])
+
+-   a function (including a class constructor or methods) argument *requires a more specific (“narrower”) type*, for example if it previously accepted `string | number` but now requires `string`—since the user will have to change some calls to the function ([playground][narrower-argument])
+
+-   a function (including a class constructor or method) *returns a less specific (“wider”) type*, for example if it previously returned `string` but now returns `string | null`—since the user’s existing handling of the return value will be wrong in some cases ([playground][wider-return])
+
+-   a function (including a class constructor or method) adds any new *required* arguments—since all user invocations of the function will now be broken ([playground][new-required-argument])
+
+-   a function (including a class constructor or method) removes an existing argument entirely—since all user invocations of the function will now fail to type-check ([playground][remove-argument])
+
+-   a `readonly` object property type becomes a *less specific (“wider”) type*, for example if it was previously `string` but now is `string | string[]`—since the user’s existing handling of the property will be wrong in some cases ([playground][wider-property]—note that the playground uses a class but an interface or type alias would have the same behavior)
+
+-   a non-`readonly` object type changes in any way, for example if it was previously `string` but now is `string | number` *or* if it was previously `string | number` but now is `string`—since some of the user’s existing *writes* to the property will now be wrong.
```

Comment:
what about changes that are backwards-compatible? for example, adding a new property to an object type

Reply:
I'll add a separate case for adding a property to an object type, and have clarified this case about changes to existing properties.